### ISSUE TITLE
Lazy snapshot restores

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#50999]
Today when users restore a snapshot Elasticsearch creates an index by copying the contents of the repository onto disk, and only starts each shard once the copy is complete. However in many cases we do not need to copy the whole shard to disk before being able to search it. This meta-issue tracks the work towards an alternative restore mode which works lazily, starting the shard up after restoring only a fraction of the data and restoring the remaining data in the background and/or as needed to serve searches.

- [x] Add plugin infrastructure (https://github.com/elastic/elasticsearch/pull/50022)
- [x] Support searching a shard that does not really exist on disk (https://github.com/elastic/elasticsearch/pull/49651)
- [x] Support restoring an index without copying any data (https://github.com/elastic/elasticsearch/pull/50239)
- [x] Add disk-based caching layer to avoid duplicate work (https://github.com/elastic/elasticsearch/pull/50693)
- [x] Associate translog with last Lucene index commit for searchable snapshots shards (#53459)
- [x] Add instrumentation (https://github.com/elastic/elasticsearch/pull/50283)
- [x] Support restoring from S3 (#51137)
- [x] Support freezing of searchable snapshots (#52653)
- [x] Support restoring from GCS (#55403)
- [x] Support restoring from Azure (#54358)
- [ ] Support other backends TBD
- [x] Automatically reallocate and redo a restore if the primary is lost (#52527)
- [x] Integrate with ILM (#50806, #53084)
- [ ] Think about how to snapshot a cluster containing searchable snapshots (e.g. expunge relevant settings?); should a restore reinstate the searchable snapshots too?
- [ ] Disable readahead on reads that bypass the cache (via `readDirectly()`)
- [ ] ~Revisit UX decisions, particularly around needing a separate repository. Can we make the laziness a property of the individual restore instead?~ superseded by #53084 
- [ ] Gracefully handle attempts to remove a repository containing mounted snapshots.
- [ ] Documentation (concepts, APIs, ...)
- [ ] HLRC integration

**Comment by DaveCTurner** [comment]
This work is largely complete and the remaining pieces are tracked elsewhere, so I'm closing this.
